src/translator/translation_model.cpp

```diff
@@ -162,11 +175,30 @@ Ptr<marian::data::CorpusBatch> TranslationModel::convertToMarianBatch(Batch &bat
   return corpusBatch;
 }
 
-void TranslationModel::translateBatch(size_t deviceId, Batch &batch) {
+void TranslationModel::translateBatch(Workspace &workspace, Batch &batch) {
+  // We're the only people accessing this workspace, it's safe to clear.
+  // Expectation is that the workspace contains things that don't require long-term storage (per batch things).
+
+  // Parameters are stored separately and hopefully initialized and kept-isolated in the graph after
+  // scorer->init(graph) in loadBackend(...).
+
+  // This allows to avoid any leaks and generate maximum room for this incoming translation on the workspace.
+  workspace.clear();
```

Comment:
@XapaJIaMnu I've commented uncommented this with multiple models concurrently calling service.translate(...) also sequentially.

What I've observed is clearing workspace has adverse effects. I end up getting garbage outputs, probably from reading a memory which has been cleared or overwritten or something.

In any case, with clear() inactive (commented), I'm able to produce valid translations (for `head -n 1000 wngt20/sources.shuf`). 

1. In case when workspace is not cleared:
```
$ diff model_0.{orig,threaded}.txt -y --suppress-common-lines | wc -l
39
$ diff model_1.{orig,threaded}.txt -y --suppress-common-lines | wc -l
76
```

2. When workspace is cleared:
```
$ diff model_0.{orig,threaded}.txt -y --suppress-common-lines | wc -l
644
$ diff model_1.{orig,threaded}.txt -y --suppress-common-lines | wc -l
905

```

In both cases on isolated model setting,  I'm able to achieve valid translations (workspace cleared or not). In case of when both models are in play simultaneously, I get garbage outputs at times in the no workspace clear setting.